### PR TITLE
Use new conversions in patch

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #39017

### CODE REVIEW

**Comment by smarterclayton** [comment]
And for reminder for others reviewing we said we would add something equivalent to the CACHE_MUTATION_DETECTOR that would verify both paths (when the env var is on) generate the same resulting struct and we would enable that in all the core CI tests but disable in kubemark.

**Comment by wojtek-t** [comment]
@smarterclayton - yes sorry, I should have moved the comment here. It's basically still assigned to me, as this is one of the things that's not yet done.

**Comment by wojtek-t** [comment]
@k8s-bot etcd3 test this, issue: #IGNORE (cluster didn't start)

**Comment by wojtek-t** [comment]
@smarterclayton - this should be ready for review. PTAL

**Comment by wojtek-t** [comment]
@k8s-bot bazel test this, issue: #IGNORE

File: staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go [hunk 1]
```diff
@@ -173,8 +174,12 @@ func fromUnstructured(sv, dv reflect.Value) error {
 					dv.Set(sv.Convert(dt))
 					return nil
 				}
+				if sv.Float() == math.Trunc(sv.Float()) {
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go]: Add a test case for this

File: staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go [hunk 2]
```diff
@@ -82,7 +83,7 @@ var (
 func parseBool(key string) bool {
 	value, err := strconv.ParseBool(key)
 	if err != nil {
-		glog.Errorf("Couldn't parse %s as bool", key)
+		glog.Errorf("Couldn't parse '%s' as bool for unstructured mismatch detection", key)
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go]: utilruntime.HandleError

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 3]
```diff
@@ -594,10 +600,28 @@ func patchResource(
...
 			case types.StrategicMergePatchType:
-				originalMap, patchMap, err := strategicPatchObject(codec, currentObject, patchJS, objToUpdate, versionedObj)
+				// Since the patch is applied on versioned objects, we need to convert the
+				// current object to versioned representation first.
+				gvks, _, err := typer.ObjectKinds(versionedObj)
 				if err != nil {
 					return nil, err
 				}
+				gvk := gvks[0]
+				currentVersionedObject, err := convertor.ConvertToVersion(currentObject, gvk.GroupVersion())
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: You should pass `gvks` to ConvertToVersion

Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Is current object throwaway?  If it is, use UnsafeConvertToVersion

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 4]
```diff
@@ -594,10 +600,28 @@ func patchResource(
...
+				if err != nil {
+					return nil, err
+				}
+				versionedObjToUpdate, err := creater.New(gvk)
+				if err != nil {
+					return nil, err
+				}
+				originalMap, patchMap, err := strategicPatchObject(codec, currentVersionedObject, patchJS, versionedObjToUpdate, versionedObj)
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Have you measured these changes?  It's entirely possible ConvertToVersion here and below loses all the benefits you got from the more efficient method

**Review by smarterclayton — CHANGES_REQUESTED** [review]
Some comments and questions

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 4 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Even without unsafe conversion, the conversions should be pretty much the same what we have now.
Basically, the input/output objects here are unversioned. And the codec we are using is the "versioning" codec, which is also doing the conversion (the safe one) + marshaling/unmarshaling.

So if we change to unsafe (which unless I'm completely missing something we can do), this should be even better (+ the savings from more performant conversion to unstructured).

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 3 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: I don't understand how to pass gvks to it - it's a different type (it just doesn't compile for me).

Regarding UnsafeConvert - yes, this object is throwaway, but to do it:
- I would need access to runtime.Scheme (which I don't have now) or extend Convertor interface (both are not super easy)
- it shouldn't be worse that what we have now - see my comment below).

So I just dropped a TODO for it. I can work on it in a subsequent PR (though it probably won't hit 1.6).

File: staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go (hunk 2 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go]: done

File: staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go (hunk 1 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/conversion/unstructured/converter.go]: Done (unfortunately I should have done it explicitly, to force the map[string]interface{] to really contain float).

**Review by wojtek-t — COMMENTED** [review]
@smarterclayton - PTAL

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 4 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: It seems I was wrong and versioning Codec is using unsafe conversions...

Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: OK - so I managed to changed it to use unsafe conversions, which should pretty much be the same as before (when serializing/deserializing to json).

**Comment by wojtek-t** [comment]
@k8s-bot gci gce test this, issue: #IGNORE

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 5]
```diff
@@ -647,6 +649,8 @@ func patchResource(
 				return nil, err
 			}
 			gvk := gvks[0]
+			// TODO: Since <currentObject> is throwaway, we can use unsafe conversion.
+			// However, to do that, we need to pass runtime.Scheme here.
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: We can pass an unsafe convertor in on the request.Scope.  But this is fine.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 6]
```diff
@@ -718,10 +724,18 @@ func patchResource(
...
 			}
+			// Convert the object back to unversioned.
 			objToUpdate := patcher.New()
-			if err := convertor.Convert(versionedObjToUpdate, objToUpdate, nil); err != nil {
+			gvks, _, err = typer.ObjectKinds(objToUpdate)
+			if err != nil {
+				return nil, err
+			}
+			gvk = gvks[0]
+			unversionedObjToUpdate, err := unsafeConvertor.ConvertToVersion(versionedObjToUpdate, gvk.GroupVersion())
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Pass `gvks` here

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 7]
```diff
@@ -649,9 +657,7 @@ func patchResource(
 				return nil, err
 			}
 			gvk := gvks[0]
-			// TODO: Since <currentObject> is throwaway, we can use unsafe conversion.
-			// However, to do that, we need to pass runtime.Scheme here.
-			currentVersionedObject, err := convertor.ConvertToVersion(currentObject, gvk.GroupVersion())
+			currentVersionedObject, err := unsafeConvertor.ConvertToVersion(currentObject, gvk.GroupVersion())
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Pass `gvks` here

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 8]
```diff
@@ -621,9 +620,18 @@ func patchResource(
...
 					return nil, err
 				}
-				if err := convertor.Convert(versionedObjToUpdate, objToUpdate, nil); err != nil {
+				// Convert the object back to unversioned.
+				gvks, _, err = typer.ObjectKinds(objToUpdate)
+				if err != nil {
+					return nil, err
+				}
+				gvk = gvks[0]
+				unversionedObjToUpdate, err := unsafeConvertor.ConvertToVersion(versionedObjToUpdate, gvk.GroupVersion())
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: `gvks`

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 9]
```diff
@@ -607,9 +608,7 @@ func patchResource(
 					return nil, err
 				}
 				gvk := gvks[0]
-				// TODO: Since <currentObject> is throwaway, we can use unsafe conversion.
-				// However, to do that, we need to pass runtime.Scheme here.
-				currentVersionedObject, err := convertor.ConvertToVersion(currentObject, gvk.GroupVersion())
+				currentVersionedObject, err := unsafeConvertor.ConvertToVersion(currentObject, gvk.GroupVersion())
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: `gvks`

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 5 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Fixed in the next commit.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 6 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: I don't understand it - it doesn't compile for me.
[]GVK doesn't implement GroupVersioner.

**Review by smarterclayton — APPROVED** [review]

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 6 above)
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Discussed offline - added TODO (in all places).

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 10]
```diff
@@ -594,10 +601,37 @@ func patchResource(
...
 			case types.StrategicMergePatchType:
-				originalMap, patchMap, err := strategicPatchObject(codec, currentObject, patchJS, objToUpdate, versionedObj)
+				// Since the patch is applied on versioned objects, we need to convert the
+				// current object to versioned representation first.
+				gvks, _, err := typer.ObjectKinds(versionedObj)
 				if err != nil {
 					return nil, err
 				}
+				gvk := gvks[0]
+				currentVersionedObject, err := unsafeConvertor.ConvertToVersion(currentObject, gvk.GroupVersion())
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Hrm, I was wrong - you need: `scope.Kind.GroupVersion()` here

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 11]
```diff
@@ -594,10 +601,37 @@ func patchResource(
...
+				if err != nil {
+					return nil, err
+				}
+				versionedObjToUpdate, err := creater.New(gvk)
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: scope.Kind

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go (hunk 8 above)
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: Oops, this should be scope.Kind.GroupVersion()

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 12]
```diff
@@ -84,14 +85,8 @@ func strategicPatchObject(
...
 ) (originalObjMap map[string]interface{}, patchMap map[string]interface{}, retErr error) {
-	// TODO: This should be one-step conversion that doesn't require
-	// json marshaling and unmarshaling once #39017 is fixed.
-	data, err := runtime.Encode(codec, originalObject)
-	if err != nil {
-		return nil, nil, err
-	}
 	originalObjMap = make(map[string]interface{})
-	if err := json.Unmarshal(data, &originalObjMap); err != nil {
+	if err := unstructured.DefaultConverter.ToUnstructured(originalObject, &originalObjMap); err != nil {
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: I'd like the comment on this method to indicate that this expects a versioned object, not the internal object like it did before.  Also, no other callers in this package?

Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: Added comment. And yes - this is the only caller.

**Comment by wojtek-t** [comment]
@smarterclayton - PTAL

**Pull request merged**